pkg/cmd/server/origin/master.go

```diff
@@ -257,6 +258,16 @@ func (c *MasterConfig) Run(kubeAPIServerConfig *kubeapiserver.Config, controller
 	// add post-start hooks
 	aggregatedAPIServer.GenericAPIServer.AddPostStartHookOrDie("template.openshift.io-sharednamespace", c.ensureOpenShiftSharedResourcesNamespace)
 	aggregatedAPIServer.GenericAPIServer.AddPostStartHookOrDie("authorization.openshift.io-bootstrapclusterroles", bootstrappolicy.Policy().EnsureRBACPolicy())
+	aggregatedAPIServer.GenericAPIServer.AddPostStartHookOrDie("admission.openshift.io-RefreshRESTMapper", func(context apiserver.PostStartHookContext) error {
+		c.RESTMapper.Reset()
+		go func() {
+			wait.PollUntil(10*time.Second, func() (bool, error) {
```

Comment:
I don't have the upstream issues handy where @caesarxuchao noted the performance characteristics of this refresh, but I seem to recall it not being great. Would it be worth at least making the period configurable?

Reply:
> I don't have the upstream issues handy where @caesarxuchao noted the performance characteristics of this refresh, but I seem to recall it not being great. Would it be worth at least making the period configurable?

We'll get to see how bad it is.  Without rate limiting, I would expect the calls to be quick for discovery.

Reply:
how will performance issues here surface? 10 seconds seems aggressive to me. also, does use of this this tolerate partial discovery failures?

Reply:
Do you need to catch panics since this is in a different go routine?

Reply:
> 10 seconds seems aggressive to me

And arbitrary. 

> does use of this this tolerate partial discovery failures?

Since we blindly reset the rest mapper, it would be possible to get into a state where `GetAPIGroupResources` fails and we have no longer have the old rest mapper.

Reply:
> how will performance issues here surface? 10 seconds seems aggressive to me. also, does use of this this tolerate partial discovery failures?

Better hope so, since its upstream.

Reply:
> 10 seconds seems aggressive to me. also, does use of this this tolerate partial discovery failures?

Got a preferred time?  I don't like waiting, so I tend to prefer tighter times.



> And arbitrary.

Any choice would be arbitrary.

Reply:
> Since we blindly reset the rest mapper, it would be possible to get into a state where GetAPIGroupResources fails and we have no longer have the old rest mapper.

Doesn't look like it to me.  @ironcladlou's implementation does a refresh on invalidate.  If you can't make contact, you keep using the old results.  Individual servers which fail are skipped and old results are used if they are present.



> how will performance issues here surface?

On the invalidating gofunc  (not the admission gofunc), you get lag as it waits.  Old results are used for admission.  It all seems pretty reasonable to me.

Reply:
> Do you need to catch panics since this is in a different go routine?

I was reasonably sure that the `wait.*` functions protected the callers, but I'm having trouble finding it.